src/internal/connector/exchange/container_resolver.go

```diff
@@ -51,38 +51,48 @@ type containerResolver struct {
 func (cr *containerResolver) IDToPath(
 	ctx context.Context,
 	folderID string,
-) (*path.Builder, error) {
-	return cr.idToPath(ctx, folderID, 0)
+	useIDInPath bool,
+) (*path.Builder, *path.Builder, error) {
+	return cr.idToPath(ctx, folderID, 0, useIDInPath)
 }
 
 func (cr *containerResolver) idToPath(
 	ctx context.Context,
 	folderID string,
 	depth int,
-) (*path.Builder, error) {
+	useIDInPath bool,
+) (*path.Builder, *path.Builder, error) {
 	if depth >= maxIterations {
-		return nil, errors.New("path contains cycle or is too tall")
+		return nil, nil, errors.New("path contains cycle or is too tall")
 	}
 
 	c, ok := cr.cache[folderID]
 	if !ok {
-		return nil, errors.Errorf("folder %s not cached", folderID)
+		return nil, nil, errors.Errorf("folder %s not cached", folderID)
 	}
 
 	p := c.Path()
 	if p != nil {
-		return p, nil
+		return p, c.Location(), nil
 	}
 
-	parentPath, err := cr.idToPath(ctx, *c.GetParentFolderId(), depth+1)
+	parentPath, parentLoc, err := cr.idToPath(ctx, *c.GetParentFolderId(), depth+1, useIDInPath)
 	if err != nil {
-		return nil, errors.Wrap(err, "retrieving parent folder")
+		return nil, nil, errors.Wrap(err, "retrieving parent folder")
 	}
 
-	fullPath := parentPath.Append(*c.GetDisplayName())
+	toAppend := *c.GetDisplayName()
+	if useIDInPath {
+		toAppend = *c.GetId()
+	}
+
+	fullPath := parentPath.Append(toAppend)
+	locPath := parentLoc.Append(*c.GetDisplayName())
```

Comment:
do we need a nil check around this call? I honestly don't remember if this is guaranteed to be non-nil at this point

Reply:
oof, yeah.